File: app/src/main/java/com/mdmobile/cyclops/services/AccountAuthenticator.java
package com.mdmobile.cyclops.services;


import android.accounts.AbstractAccountAuthenticator;
import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.AccountManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.Nullable;

import android.util.Log;

import com.android.volley.VolleyError;
import com.mdmobile.cyclops.api.ApiRequestManager;
import com.mdmobile.cyclops.dataModel.Server;
import com.mdmobile.cyclops.dataModel.api.Token;
import com.mdmobile.cyclops.interfaces.NetworkCallBack;
import com.mdmobile.cyclops.security.ServerNotFound;
import com.mdmobile.cyclops.ui.logIn.LoginActivity;
import com.mdmobile.cyclops.utils.Logger;
import com.mdmobile.cyclops.utils.ServerUtility;
import com.mdmobile.cyclops.utils.UserUtility;

import javax.net.ssl.HttpsURLConnection;

public class AccountAuthenticator extends AbstractAccountAuthenticator {

    //Authenticator intent keys
    public final static String ACCOUNT_TYPE_KEY = "AccountTypeKey";
    public final static String AUTH_TOKEN_TYPE_KEY = "AuthTokenTypeKey";
    public final static String AUTH_TOKEN_EXPIRATION_KEY = "AuthTokenExpirationKey";
    public final static String REFRESH_AUTH_TOKEN_KEY = "RefreshAuthTokenKey";
    public final static String ADDING_NEW_ACCOUNT_KEY = "AddingNewAccountIntentKey";
    private Context mContext;
    private String LOG_TAG = AccountAuthenticator.class.getSimpleName();

    AccountAuthenticator(Context context) {
        super(context);
        mContext = context.getApplicationContext();
    }


    @Override
    public Bundle editProperties(AccountAuthenticatorResponse accountAuthenticatorResponse, String s) {
        throw new UnsupportedOperationException("Confirm credentials is not a supported operation");

    }

    @Override
    public Bundle addAccount(AccountAuthenticatorResponse response, String accountType, String authTokenType,
                             String[] requiredFeatures, Bundle options) {

        return promptError(response, accountType, authTokenType, true);
    }

    @Override
    public Bundle confirmCredentials(AccountAuthenticatorResponse accountAuthenticatorResponse,
                                     Account account, Bundle bundle) {
        throw new UnsupportedOperationException("Confirm credentials is not a supported operation");
    }

    @Override
    public Bundle getAuthToken(final AccountAuthenticatorResponse authenticatorResponse, final Account account,
                               final String authTokenType, Bundle bundle) {


        final AccountManager accountManager = AccountManager.get(mContext);
        final Bundle userInfo = UserUtility.getUserInfo(account);
        final Server serverInfo;
        try {
            serverInfo = ServerUtility.getActiveServer();
        } catch (ServerNotFound e) {
            e.printStackTrace();
            return null;
        }

        if (userInfo == null) {
            return promptError(authenticatorResponse, account.type, authTokenType, null);
        }

        Logger.log(LOG_TAG, "Requesting new token...", Log.VERBOSE);

        ApiRequestManager.getInstance()
                .getToken(serverInfo, account.name, accountManager.getPassword(account),
                        new NetworkCallBack() {
                            @Override
                            public void tokenReceived(Bundle userInfo, Token JsonToken) {
                                Bundle result = new Bundle();
                                result.putString(AccountManager.KEY_ACCOUNT_NAME, account.name);
                                result.putString(AccountManager.KEY_ACCOUNT_TYPE, account.type);
                                result.putString(AUTH_TOKEN_TYPE_KEY, JsonToken.getToken_type());
                                result.putString(AccountManager.KEY_AUTHTOKEN, JsonToken.getAccess_token());
                                authenticatorResponse.onResult(result);
                            }

                            @Override
                            public void errorReceivingToken(VolleyError errorResponse) {
                                //If we are here with error 400 it only means credentials have changed
                                //Return the error to authenticatorResponse
                                if (errorResponse.networkResponse.statusCode == HttpsURLConnection.HTTP_UNAUTHORIZED
                                        || errorResponse.networkResponse.statusCode == HttpsURLConnection.HTTP_BAD_REQUEST) {
                                    Bundle result = promptError(authenticatorResponse, account.type, authTokenType, null);
                                    authenticatorResponse.onResult(result);
                                }
                            }
                        });
        return null;
    }

    @Override
    public String getAuthTokenLabel(String s) {
        return null;
    }

    @Override
    public Bundle updateCredentials(AccountAuthenticatorResponse accountAuthenticatorResponse,
                                    Account account, String s, Bundle bundle) {
        return null;
    }

    @Override
    public Bundle hasFeatures(AccountAuthenticatorResponse accountAuthenticatorResponse,
                              Account account, String[] strings) {
        return null;
    }


    //Utility method to return a bundle with KEY_INTENT key and the intent to show login activity
    private Bundle promptError(AccountAuthenticatorResponse response, String accountType,
                               String authTokenType, @Nullable Boolean addingNewAccount) {
        final Intent intent = new Intent(mContext, LoginActivity.class);

        //Insert the parameter required from LoginActivity to create a new account
        intent.putExtra(ACCOUNT_TYPE_KEY, accountType);
        intent.putExtra(AUTH_TOKEN_TYPE_KEY, authTokenType);
        if (addingNewAccount != null) {
            intent.putExtra(ADDING_NEW_ACCOUNT_KEY, addingNewAccount);
        }
        intent.putExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE, response);
        Bundle bundle = new Bundle();
        bundle.putParcelable(AccountManager.KEY_INTENT, intent);
        return bundle;
    }
}

File: app/src/main/java/com/mdmobile/cyclops/services/AccountAuthenticator.kt
package com.mdmobile.cyclops.services


import android.accounts.AbstractAccountAuthenticator
import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.content.Context
import android.content.Intent
import android.os.Bundle

import android.util.Log

import com.android.volley.VolleyError
import com.mdmobile.cyclops.api.ApiRequestManager
import com.mdmobile.cyclops.dataModel.Instance
import com.mdmobile.cyclops.dataModel.api.Token
import com.mdmobile.cyclops.interfaces.NetworkCallBack
import com.mdmobile.cyclops.security.ServerNotFound
import com.mdmobile.cyclops.ui.logIn.LoginActivity
import com.mdmobile.cyclops.util.Logger
import com.mdmobile.cyclops.util.ServerUtility
import com.mdmobile.cyclops.util.UserUtility

import javax.net.ssl.HttpsURLConnection

class AccountAuthenticator internal constructor(context: Context) : AbstractAccountAuthenticator(context) {
    private val mContext: Context = context.applicationContext
    private val logTag = AccountAuthenticator::class.java.simpleName


    override fun editProperties(accountAuthenticatorResponse: AccountAuthenticatorResponse, s: String): Bundle {
        throw UnsupportedOperationException("Confirm credentials is not a supported operation")

    }

    override fun addAccount(response: AccountAuthenticatorResponse, accountType: String, authTokenType: String,
                            requiredFeatures: Array<String>, options: Bundle): Bundle {

        return promptError(response, accountType, authTokenType, true)
    }

    override fun confirmCredentials(accountAuthenticatorResponse: AccountAuthenticatorResponse,
                                    account: Account, bundle: Bundle): Bundle {
        throw UnsupportedOperationException("Confirm credentials is not a supported operation")
    }

    override fun getAuthToken(authenticatorResponse: AccountAuthenticatorResponse, account: Account,
                              authTokenType: String, bundle: Bundle): Bundle? {


        val accountManager = AccountManager.get(mContext)
        val userInfo = UserUtility.getUserInfo(account)
        val instanceInfo: Instance
        try {
            instanceInfo = ServerUtility.getActiveServer()
        } catch (e: ServerNotFound) {
            e.printStackTrace()
            return null
        }

        if (userInfo == null) {
            return promptError(authenticatorResponse, account.type, authTokenType, null)
        }

        Logger.log(logTag, "Requesting new _token...", Log.VERBOSE)

        ApiRequestManager.getInstance()
                .getToken(instanceInfo, account.name, accountManager.getPassword(account),
                        object : NetworkCallBack {
                            override fun tokenReceived(userInfo: Bundle, JsonToken: Token) {
                                val result = Bundle()
                                result.putString(AccountManager.KEY_ACCOUNT_NAME, account.name)
                                result.putString(AccountManager.KEY_ACCOUNT_TYPE, account.type)
                                result.putString(AUTH_TOKEN_TYPE_KEY, JsonToken.token_type)
                                result.putString(AccountManager.KEY_AUTHTOKEN, JsonToken.access_token)
                                authenticatorResponse.onResult(result)
                            }

                            override fun errorReceivingToken(errorResponse: VolleyError) {
                                //If we are here with error 400 it only means credentials have changed
                                //Return the error to authenticatorResponse
                                if (errorResponse.networkResponse.statusCode == HttpsURLConnection.HTTP_UNAUTHORIZED || errorResponse.networkResponse.statusCode == HttpsURLConnection.HTTP_BAD_REQUEST) {
                                    val result = promptError(authenticatorResponse, account.type, authTokenType, null)
                                    authenticatorResponse.onResult(result)
                                }
                            }
                        })
        return null
    }

    override fun getAuthTokenLabel(s: String): String? {
        return null
    }

    override fun updateCredentials(accountAuthenticatorResponse: AccountAuthenticatorResponse,
                                   account: Account, s: String, bundle: Bundle): Bundle? {
        return null
    }

    override fun hasFeatures(accountAuthenticatorResponse: AccountAuthenticatorResponse,
                             account: Account, strings: Array<String>): Bundle? {
        return null
    }


    //Utility method to return a bundle with KEY_INTENT key and the intent to show login activity
    private fun promptError(response: AccountAuthenticatorResponse, accountType: String,
                            authTokenType: String, addingNewAccount: Boolean?): Bundle {
        val intent = Intent(mContext, LoginActivity::class.java)

        //Required for login activity to create a new account
        intent.putExtra(ACCOUNT_TYPE_KEY, accountType)
        intent.putExtra(AUTH_TOKEN_TYPE_KEY, authTokenType)

        if (addingNewAccount != null) {
            intent.putExtra(ADDING_NEW_ACCOUNT_KEY, addingNewAccount)
        }
        intent.putExtra(AccountManager.KEY_ACCOUNT_AUTHENTICATOR_RESPONSE, response)
        val bundle = Bundle()
        bundle.putParcelable(AccountManager.KEY_INTENT, intent)
        return bundle
    }

    companion object {
        const val REFRESH_AUTH_TOKEN_KEY = "RefreshAuthTokenKey"
        const val ADDING_NEW_ACCOUNT_KEY = "AddingNewAccountIntentKey"
        const val ACCOUNT_TYPE_KEY = "AccountTypeKey"
        const val AUTH_TOKEN_TYPE_KEY = "AuthTokenTypeKey"
        const val AUTH_TOKEN_EXPIRATION_KEY = "AuthTokenExpirationKey"
    }
}
